src/fsharp/IlxGen.fs

```diff
@@ -3066,34 +3176,43 @@ and GenApp cenv cgbuf eenv (f, fty, tyargs, curriedArgs, m) sequel =
             // when branch-calling methods we must have the right type parameters
             (match kind with
              | BranchCallClosure _ -> true
-             | BranchCallMethod (_, _, tps, _, _) ->
+             | BranchCallMethod (_, _, tps, _, _, _) ->
                   (List.lengthsEqAndForall2 (fun ty tp -> typeEquiv g ty (mkTyparTy tp)) tyargs tps)) &&
             // must be exact #args, ignoring tupling - we untuple if needed below
             (let arityInfo =
                match kind with
                | BranchCallClosure arityInfo
-               | BranchCallMethod (arityInfo, _, _, _, _) -> arityInfo
+               | BranchCallMethod (arityInfo, _, _, _, _, _)  -> arityInfo
              arityInfo.Length = curriedArgs.Length
             ) &&
             (* no tailcall out of exception handler, etc. *)
             (match sequelIgnoringEndScopesAndDiscard sequel with Return | ReturnVoid -> true | _ -> false))
         | None -> false
     ->
         let (kind, mark) = ListAssoc.find g.valRefEq v eenv.innerVals // already checked above in when guard
+
+        // Generate the arguments for the direct tail call.
+        // We push all the arguments on the IL stack then write them back to the argument slots using
+        // I_starg.  This seems a little sloppy, we could generate-then-write for each of the arguments.
+        //
+        // The arguments pushed don't include the 'this' argument for a recursive closure call (in PreallocatedArgCount)
+        // The arguments _do_ include the 'this' argument for instance method calls.  The arguments do _not_ include witness arguments.
         match kind with
         | BranchCallClosure arityInfo ->
             GenExprs cenv cgbuf eenv curriedArgs
 
             let numArgs = List.sum arityInfo
 
+            // TODO: witness argument generation for closures
```

Comment:
Will doing the work for this "TODO" later be another language feature and/or break compat?

Reply:
Just asking, as I'm having a bit of a time understanding this whole bit in "GenApp".

Reply:
This TODO is not real - I have removed it.  Closures capture sitnesses but do not take them as arguments at their invocation methods

I think we should clean up GenApp and separate its branches